### PR TITLE
By reference enum param fixes

### PR DESCRIPTION
**Pull request by IIFE**
- Enum passed by reference is getting treated as a raw pointer, where enum is being dereferenced when it shouldn't.
- Full qualified enum name is not being used when generating enum param.

### CODE REVIEW

File: src/Generator/Generators/CLI/CLITypePrinter.cs [hunk 1]
```diff
@@ -146,11 +146,11 @@ public override TypePrinterResult VisitDelegate(FunctionType function)
             Enumeration @enum;
             if (pointee.TryGetEnum(out @enum))
             {
-                var typeName = @enum.Visit(this);
+                var typeName = VisitDeclaration(@enum, quals);
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLITypePrinter.cs:149]: What's the reason for this change?

File: src/Generator/Generators/CLI/CLITypePrinter.cs [hunk 2]
```diff
@@ -146,11 +146,11 @@ public override TypePrinterResult VisitDelegate(FunctionType function)
...
 
                 // Skip one indirection if passed by reference
-                if (Parameter != null && (Parameter.IsOut || Parameter.IsInOut)
-                    && pointee == finalPointee)
+                if (Parameter != null && (Parameter.Type.IsReference() || ((Parameter.IsOut || Parameter.IsInOut)
```
Comment (ddobrev) [inline src/Generator/Generators/CLI/CLITypePrinter.cs:152]: This line is now too long, you could break it at `((Parameter.IsOut`.

File: tests/CLI/CLI.h [hunk 3]
```diff
@@ -55,4 +55,16 @@ struct CompleteIncompleteStruct;
 typedef struct IncompleteStruct IncompleteStruct;
 
 DLL_API IncompleteStruct* createIncompleteStruct();
-DLL_API void useIncompleteStruct(IncompleteStruct* a);
\ No newline at end of file
+DLL_API void useIncompleteStruct(IncompleteStruct* a);
```
Comment (ddobrev) [inline tests/CLI/CLI.h:58]: We try placing non-generator specific tests in `Common.h`. Could you please try moving it there?

File: src/Generator/Generators/CLI/CLITypePrinter.cs (hunk 1 above)
Comment (IIFE) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: The generator does not generate fully qualified parameter name. For example instead of `CLI::EnumParam` it just generates `EnumParam`, which does not build.

File: tests/CLI/CLI.h (hunk 3 above)
Comment (IIFE) [reply tests/CLI/CLI.h:58]: The fix is specific to CLI because it is fixing pointer dereferencing.

File: src/Generator/Generators/CLI/CLITypePrinter.cs (hunk 1 above)
Comment (ddobrev) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: We should perhaps fix the real issue, that is, make visiting an enum generate the qualified name. @tritao what do you think?

File: tests/CLI/CLI.h (hunk 3 above)
Comment (ddobrev) [reply tests/CLI/CLI.h:58]: @IIFE I meant that functions taking a pointer to an enum can be mapped to C# as well. So please try moving the test there as I expect no trouble from the C# gen, it can already handle such pointers.

File: src/Generator/Generators/CLI/CLITypePrinter.cs (hunk 1 above)
Comment (IIFE) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: It does generate qualified name for other enum scenarios, does not seem to do it if enum is passed by ref, hence this fix.

Comment (tritao) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: You should be able to do `@enum.Visit(this, quals);`.

Can you give that a try?

Comment (IIFE) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: There is no `Visit` overload that takes a second quals param.

Comment (tritao) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: You're right, it's only supported for types, my bad.

Comment (tritao) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: @IIFE Can you try removing the following method from `CLITypePrinter`?

```csharp
        public override TypePrinterResult VisitEnumDecl(Enumeration @enum)
        {
            return @enum.Name;
        }
``` 

That will call into the base `CppTypePrinter`, which should print enums using the general qualified declaration code.

**Comment by tritao** [comment]
This looks mostly good, I would just like for you to try my suggestion to use the base printer for enums. And if you could move the test to `Common`, as @ddobrev suggested, we'd definitely prefer as that increases test coverage of other backends aswell.

File: tests/CLI/CLI.h (hunk 3 above)
Comment (IIFE) [reply tests/CLI/CLI.h:58]: Just tried common test. Unfortunately the C# generation also generates pointer to by ref enum param, so moving the test to common will break unless the C# part is fixed, which is probably worth doing in another PR, so maybe keep the CLI tests?

File: src/Generator/Generators/CLI/CLITypePrinter.cs (hunk 1 above)
Comment (IIFE) [reply src/Generator/Generators/CLI/CLITypePrinter.cs:149]: Removing that method causes StackOverflowException when generator is run.

File: src/Generator/Generators/CLI/CLITypePrinter.cs (hunk 2 above)
Comment (IIFE) [reply src/Generator/Generators/CLI/CLITypePrinter.cs]: Done

File: tests/CLI/CLI.h (hunk 3 above)
Comment (tritao) [reply tests/CLI/CLI.h:58]: Thanks for checking, we can fix it later.
